packages/graphback-datasync/src/providers/DeltaDbDataProvider.ts

```diff
@@ -0,0 +1,178 @@
+import { GraphQLObjectType } from 'graphql';
+import { getDatabaseArguments, metadataMap, GraphbackContext, NoDataError, TransformType, FieldTransform, GraphbackOrderBy, GraphbackPage } from '@graphback/core';
+import { ObjectId } from 'mongodb';
+import { MongoDBDataProvider, applyIndexes } from '@graphback/runtime-mongo';
+import { ConflictError, ConflictStateMap } from "../util";
+import { DataSyncProvider } from "./DataSyncProvider";
+import { DeltaProvider } from "./DeltaProvider";
+
+/**
+ * Mongo provider that attains data synchronization using soft deletes
+ */
+export class DeltaDBDataProvider<Type = any> extends MongoDBDataProvider<Type> implements DataSyncProvider {
+  protected deltaProvider: DeltaProvider;
+
+  public constructor(baseType: GraphQLObjectType, client: any) {
+    super(baseType, client);
+    applyIndexes([
+      {
+        key: {
+          _deleted: 1
+        }
+      }
+    ], this.db.collection(this.collectionName)).catch((e: any) => {
+      throw e;
+    });
+    this.deltaProvider = new DeltaProvider(baseType, client);
+  }
+
+  public async create(data: any, context: GraphbackContext): Promise<Type> {
+    data._deleted = false;
+
+    const res = await super.create(data, context);
+
+    this.deltaProvider.insertDiff(res).catch((e: any) => {
+      console.error(`Error: Couldn't insert into delta table: ${e}`);
+    })
+    
+    return this.mapFields(res);
+  }
+
+  public async update(data: any, context: GraphbackContext): Promise<Type> {
+    const conflict = await this.checkForConflicts(data, context);
+
+    if (conflict !== undefined) {
+      throw new ConflictError(conflict);
+    }
+
+    // TODO use findOneAndUpdate to check consistency afterwards
+    const { idField } = getDatabaseArguments(this.tableMap, data);
+
+    if (!idField.value) {
+      throw new NoDataError(`Cannot update ${this.collectionName} - missing ID field`)
+    }
+
+    this.fieldTransformMap[TransformType.UPDATE]
+      .forEach((f: FieldTransform) => {
+        data[f.fieldName] = f.transform(f.fieldName);
+      });
+
+    const objectId = new ObjectId(idField.value);
+    const result = await this.db.collection(this.collectionName).findOneAndUpdate({ _id: objectId }, { $set: data }, { returnOriginal: false });
+    if (result.ok) {
+      
+      this.deltaProvider.insertDiff(result.value).catch((e: any) => {
+        console.error(`Error: Couldn't insert into delta table: ${e}`);
+      })
```

Comment:
This block in `update`, `delete` and `create` seem to be the only things that make this provider different to `DatasyncMongoDBDataProvider`.

Could this just extend that instead, to save repeating code in two providers and increasing maintenance costs.

Or would it be feasible to use the same provider, and only execute this block if `this.deltaProvider` has been created, which could be conditionally created based on the existence of the the `deltaTable` annotation key.

This got me thinking also - maybe we should be passing the `ModelDefinition` to the provider instead of the base GraphQL type, as a lot of the preparsed metadata information is not available from the base type and you would have to re-parse it.

Reply:
> This got me thinking also - maybe we should be passing the ModelDefinition to the provider instead of the base GraphQL type, as a lot of the preparsed metadata information is not available from the base type and you would have to re-parse it.

+1, this rejoins this comment https://github.com/aerogear/graphback/pull/1614#discussion_r448159331